### DISCUSSION TITLE
Using Encrypted Key Files with x509 Certificates for Authentication in EMQX?

### DISCUSSION THREAD

**Discussion by valientegaston** [emqx/emqx#13937] [General]
Hi, I'm just getting started with brokers and MQTT, and I was wondering if it's possible to use x509 certificates for authentication in EMQX, but with encrypted key files. I haven't found any info on how to provide the decryption key for those certificates. Any guidance would be awesome. Thanks!

**Comment by qzhuyan** [comment]
do you mean password protected?

**Reply by valientegaston** [reply]
Yes, I meant password protected. Thank you.

**Reply by thalesmg** [reply]
You can set the password via configuration: https://docs.emqx.com/en/enterprise/v5.8.0/hocon/#V-listeners-S-listeners-ssl-S-mqtt_ssl_listener-ssl_options-S-listener_ssl_opts-password

Ex:

```hcl
listeners.ssl.default.ssl_options.password = xxx
```

Or just use the dashboard (under Cluster Settings > Listeners > default ssl > Advanced Settings).

![image](https://github.com/user-attachments/assets/aad63ea7-9dd4-4983-9acd-2931b2508d13)

**Reply by valientegaston** [reply]
Sorry, I'm trying to understand how everything works. What I want to do is use X509 certificates on clients to authenticate with the broker. I also use SSL. I have generated the CA certificates, the broker's, and the client's, all with their respective key files without a password, and everything works. My second test would be to use a password when generating the certificates and key files. As a security measure, if someone manages to extract the client's key file from the devices, the security won't be compromised if they don't have the password.

I believe the configuration you provided is for using a password with SSL certificates, but not for what I'm trying to do, is that correct? Thank you very much!

**Reply by id** [reply]
@valientegaston no, the password is for the key file. Certificates are never password protected, it's a public key.
